src/Servers/Kestrel/Core/src/Internal/ServiceContext.cs

```diff
@@ -7,22 +7,25 @@
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
 {
+    // Ideally this type should be readonly and initialized with a constructor.
+    // Tests use TestServiceContext which inherits from this type and sets properties.
+    // Changing this type would be a lot of work.
```

Comment:
The TestServiceContext sets all of these except for Heartbeat in its ctor. We could add the ctor but keep the setters.

Reply:
Meh, the only reason I've been adding constructors is because properties + nullable doesn't work well. Since everything here is forced to have a value then it doesn't matter.